File: spec/javascripts/components/error-summary-spec.js
/* global describe afterEach beforeEach it expect */

var $ = window.jQuery

function isFocused (element) {
  return $(element)[0] === $(element)[0].ownerDocument.activeElement
}

describe('An error summary component', function () {
  'use strict'

  var GOVUK = window.GOVUK
  var module
  var element

  beforeEach(function () {
    element = $(
      '<div ' +
        'class="gem-c-error-summary" ' +
        'data-module="error-summary" ' +
        'aria-labelledby="error-summary-title-id" ' +
        'role="alert" ' +
        'tabindex="-1" ' +
      '> ' +
        '<h2 class="gem-c-error-summary__title" id="error-summary-title-id"> ' +
          'Message to alert the user to a problem goes here ' +
        '</h2> ' +
        '<div class="gem-c-error-summary__body"> ' +
          '<p class="gem-c-error-summary__text"> ' +
            'Optional description of the errors and how to correct them ' +
          '</p> ' +
          '<ul class="gem-c-error-summary__list"> ' +
            '<li class="gem-c-error-summary__list__item"> ' +
              '<a ' +
                'class="js-error-summary__link gem-c-error-summary__link" ' +
                'href="#example-error-1" ' +
              '>Descriptive link to the question with an error</a> ' +
            '</li> ' +
          '</ul> ' +
        '</div> ' +
      '</div>'
    )
    $(document.body).append(element)
    document.body.focus()
    module = new GOVUK.Modules.ErrorSummary()
  })

  afterEach(function () {
    // clean up the test element after each test
    element.remove()
    element = undefined
  })

  it('is focused on load', function () {
    module.start(element)

    expect(isFocused(element)).toBe(true)
  })

  it('focuses inputs with no location change when clicking error links', function () {
    var originalHref = window.location.href
    var inputElement = $('<input type="text" id="example-error-1" name="example-error-1" />')
    $(document.body).append(inputElement)

    module.start(element)

    element.find('.js-error-summary__link').trigger('click')

    expect(isFocused(inputElement)).toBe(true)
    expect(originalHref).toBe(window.location.href)
  })
})

File: spec/components/error_summary_spec.rb
require "rails_helper"

describe "Error summary", type: :view do
  def component_name
    "error_summary"
  end

  it "fails to render when no data is given" do
    assert_raises do
      render_component({})
    end
  end

  it "renders error items when no title is set" do
    render_component(
      id: "error-summary-id",
      items: [
        {
          text: "Descriptive link to the question with an error",
          href: "#example-error-1",
        },
      ],
    )

    assert_select ".gem-c-error-summary[id='error-summary-id']"
    assert_select(
      "ul li a:first-of-type[href='#example-error-1']",
      text: "Descriptive link to the question with an error",
    )
  end

  it "renders an error summary with title set correctly" do
    render_component(
      id: "error-summary-id",
      title: "Message to alert the user to a problem goes here",
    )

    assert_select ".gem-c-error-summary[id='error-summary-id']"
    assert_select ".govuk-error-summary__title", text: "Message to alert the user to a problem goes here"
    assert_select ".govuk-error-summary__body p", count: 0
    assert_select ".govuk-error-summary__list", count: 0
  end

  it "renders the component wrapped in a div with role alert" do
    render_component(
      id: "error-summary-id",
      title: "Message to alert the user to a problem goes here",
    )

    assert_select ".gem-c-error-summary.govuk-error-summary [role='alert']"
  end

  it "renders an error summary with items links" do
    render_component(
      title: "Message to alert the user to a problem goes here",
      description: "Optional description of the errors and how to correct them",
      items: [
        {
          text: "Descriptive link to the question with an error 1",
          href: "#example-error-1",
          data_attributes: { gtm: "tracking-1" },
        },
        {
          text: "Descriptive link to the question with an error 2",
          href: "#example-error-2",
          target: "_blank",
          data_attributes: { gtm: "tracking-2" },
        },
        {
          text: "Description for error 3 with no link",
          data_attributes: { gtm: "tracking-3" },
        },
      ],
    )

    assert_select ".govuk-error-summary__title", text: "Message to alert the user to a problem goes here"
    assert_select ".govuk-error-summary__body p", text: "Optional description of the errors and how to correct them"

    assert_select(
      "ul li a:first-of-type[href='#example-error-1'][data-gtm='tracking-1']",
      text: "Descriptive link to the question with an error 1",
    )
    assert_select(
      "ul li a:nth-of-type(1)[href='#example-error-2'][data-gtm='tracking-2']",
      text: "Descriptive link to the question with an error 2",
    )
    assert_select(
      "ul li:last-of-type span[data-gtm='tracking-3']",
      text: "Description for error 3 with no link",
    )
  end
end
